ui/src/wrapApiWithVersioning.ts

```diff
@@ -0,0 +1,49 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as semver from 'semver';
+import { IActionContext } from '..';
+import { AzureExtension, AzureExtensionProvider } from '../api';
+import { callWithTelemetryAndErrorHandling } from './callWithTelemetryAndErrorHandling';
+import { getPackageInfo } from './getPackageInfo';
+import { localize } from './localize';
+
+export function wrapApiWithVersioning(azExts: AzureExtension[]): AzureExtensionProvider {
+    for (const azExt of azExts) {
+        if (!semver.valid(azExt.apiVersion)) {
+            throw new Error(localize('invalidVersion', 'Invalid semver "{0}".', azExt.apiVersion));
+        }
+    }
+
+    const extensionId: string = getPackageInfo().extensionId;
+    return {
+        getExtension: async <T extends AzureExtension>(apiVersion: string): Promise<T> => getExtensionInternal<T>(azExts, extensionId, apiVersion)
```

Comment:
Probably but it won't be simple. Specifically I want to make sure we're not changing the api in any way when we wrap it with telemetry. It might be a bit thorny figuring out what is a property vs. function vs. async function and handling each appropriately. To that end - I've added unit tests to make sure I don't break stuff when I add telemetry (preferably in a separate PR)